### DISCUSSION TITLE
Added: Notes for September 14th, 2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3892] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-14.md

## Agenda
- Partial method signature matching
- Null-conditional handling of the nullable suppression operator
- Annotating IEnumerable.Cast
- Nullability warnings in user-written record code
- Tuple deconstruction mixed assignment and declaration

**Comment by MgSam** [comment]
> Members of the LDT believe that this is broken on the same level as the for iterator variable behavior that was changed in C# 5, and we believe that we should take a similar breaking change to fix the behavior here.

I don't recall what the bug or the fix was here. Anyone else remember? Was the compiler patched prior to C# 6 to fix it?

**Reply by CyrusNajmabadi** [reply]
The issue was that if you had a foreach and you captured the foreach variable into a lambda, then as the foreach proceeded each lambda would then point at the current iteration value. 

We changed this so each lambda captures a fresh variable, so they are not impacted as the foreach proceeds.

**Reply by HaloFour** [reply]
> I don't recall what the bug or the fix was here. Anyone else remember? Was the compiler patched prior to C# 6 to fix it?

It wasn't a bug so much as it was a design decision that led to unexpected bugs.  If you captured the iterator variable in a `foreach` loop in a lambda and executed the lambda after the loop it would reflect the value at the end of the loop, not for each iteration:

```cs
var list = new List<Action>();
var numbers = Enumerable.Range(1, 10);
foreach (var number in numbers) {
    list.Add(() => Console.WriteLine(number));
}
foreach (var action in list) {
    action();
}
```

This used to print "10" 10 times.  Now it prints 1 through 10.

**Reply by MgSam** [reply]
Ah, ok, thanks. I'm aware of that one. That wasn't introduced in C# 5 but rather "fixed" in C# 5. I thought the wording made it sound like a compiler bug was shipped and patched prior to C# 6.

**Reply by 333fred** [reply]
> I thought the wording made it sound like a compiler bug was shipped and patched prior to C# 6.

No, it was a spec "feature", just like the behavior of `!`.
